### ISSUE TITLE
no-unused-vars should warn on the offending assignment instead of declaration

### ISSUE THREAD

**Issue by Jimbly** [eslint/eslint#12871]
**Tell us about your environment**

* **ESLint Version:** v6.8.0 via [eslint.org/demo](https://eslint.org/demo#eyJ0ZXh0IjoiZnVuY3Rpb24gcHJpbnQodikge1xuICBjb25zb2xlLmxvZyh2KTsgLy8gZXNsaW50LWRpc2FibGUtbGluZVxuICByZXR1cm4gMTtcbn1cbmxldCB5ID0gMTtcbnkgKz0gcHJpbnQoeSk7XG55ICs9IHByaW50KHkpO1xueSArPSBwcmludCh5KTtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnNvbGUiOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZW1wdHkiOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnsiYnJvd3NlciI6dHJ1ZSwibm9kZSI6dHJ1ZSwiY29tbW9uanMiOnRydWUsInNoYXJlZC1ub2RlLWJyb3dzZXIiOnRydWUsIndvcmtlciI6dHJ1ZSwiYW1kIjp0cnVlLCJtb2NoYSI6dHJ1ZSwiamFzbWluZSI6dHJ1ZSwiamVzdCI6dHJ1ZSwicGhhbnRvbWpzIjp0cnVlLCJqcXVlcnkiOnRydWUsInF1bml0Ijp0cnVlLCJwcm90b3R5cGVqcyI6dHJ1ZSwic2hlbGxqcyI6dHJ1ZSwibWV0ZW9yIjp0cnVlLCJtb25nbyI6dHJ1ZSwicHJvdHJhY3RvciI6dHJ1ZSwiYXBwbGVzY3JpcHQiOnRydWUsIm5hc2hvcm4iOnRydWUsInNlcnZpY2V3b3JrZXIiOnRydWUsImF0b210ZXN0Ijp0cnVlLCJlbWJlcnRlc3QiOnRydWUsIndlYmV4dGVuc2lvbnMiOnRydWUsImVzNiI6dHJ1ZSwiZ3JlYXNlbW9ua2V5Ijp0cnVlfX19)

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
default

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**
Simple example (from some actual immediate-mode UI code that takes in a y position and returns a height of what was rendered).

```js
function print(v) {
  console.log(v);
  return 1;
}
let y = 1;
y += print(y);
y += print(y);
y += print(y);
```

[eslint.org/demo#...](https://eslint.org/demo#eyJ0ZXh0IjoiZnVuY3Rpb24gcHJpbnQodikge1xuICBjb25zb2xlLmxvZyh2KTsgLy8gZXNsaW50LWRpc2FibGUtbGluZVxuICByZXR1cm4gMTtcbn1cbmxldCB5ID0gMTtcbnkgKz0gcHJpbnQoeSk7XG55ICs9IHByaW50KHkpO1xueSArPSBwcmludCh5KTtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnNvbGUiOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZW1wdHkiOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnsiYnJvd3NlciI6dHJ1ZSwibm9kZSI6dHJ1ZSwiY29tbW9uanMiOnRydWUsInNoYXJlZC1ub2RlLWJyb3dzZXIiOnRydWUsIndvcmtlciI6dHJ1ZSwiYW1kIjp0cnVlLCJtb2NoYSI6dHJ1ZSwiamFzbWluZSI6dHJ1ZSwiamVzdCI6dHJ1ZSwicGhhbnRvbWpzIjp0cnVlLCJqcXVlcnkiOnRydWUsInF1bml0Ijp0cnVlLCJwcm90b3R5cGVqcyI6dHJ1ZSwic2hlbGxqcyI6dHJ1ZSwibWV0ZW9yIjp0cnVlLCJtb25nbyI6dHJ1ZSwicHJvdHJhY3RvciI6dHJ1ZSwiYXBwbGVzY3JpcHQiOnRydWUsIm5hc2hvcm4iOnRydWUsInNlcnZpY2V3b3JrZXIiOnRydWUsImF0b210ZXN0Ijp0cnVlLCJlbWJlcnRlc3QiOnRydWUsIndlYmV4dGVuc2lvbnMiOnRydWUsImVzNiI6dHJ1ZSwiZ3JlYXNlbW9ua2V5Ijp0cnVlfX19)

**What did you expect to happen?**
No error, or, potentially an error on the first unused assignment:
`8:1 - 'y' is assigned a value but never used. (no-unused-vars)`

**What actually happened? Please include the actual, raw output from ESLint.**
`5:5 - 'y' is assigned a value but never used. (no-unused-vars)`

**Are you willing to submit a pull request to fix this bug?**
Probably not, but potentially.

Edits: Title was previously "no-unused-vars erroneously fails if a line both uses and modifies a value via increment", and I was confused by the error being on the declaration of `y` instead of the last (actually unused) assignment.

**Comment by gabrieldrs** [comment]
Although it reports the error at line 5, the actual error seems to occur because in your last line, you assign a value to y but don't use it, if you replace your last line with `print(y);` the error goes away

**Comment by Jimbly** [comment]
Ah, you are right!  I was tricked by line 5 being highlighted.  So, I guess, much more minor bug/improvement is that the line number for `no-unused-vars` should perhaps be the erroneous assignment instead of the first assignment/declaration?

**Comment by kaicataldo** [comment]
I think it makes sense to warn on the last assignment. This sounds like a good enhancement to me👍🏼

**Comment by Jimbly** [comment]
Sounds good, I've updated the title and original description to be more correct.

**Comment by gabrieldrs** [comment]
I took a look at the implementation and it seems like this behaviour is intentional (https://github.com/eslint/eslint/blob/master/lib/rules/no-unused-vars.js#L594)

I also think that the warning should report the last assignment, and I can work on this if we agree that this should be changed

**Comment by Jimbly** [comment]
I do not necessarily think it should report the last assignment, but I think it should report the first *unused* assignment, e.g., 
```javascript
var a = 1;
a = 1;
```
I think reporting the first line would be appropriate (to fix, you'd want to remove that var statement and then fix all of the "a is undefined" errors that result).

That being said, doing a little poking, and there definitely is something weird/inconsistent going on with lines that both use and modify a variable - [this](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZ2xvYmFscyBmICovXG5sZXQgeSA9IDE7XG5mKHkpO1xueSA9IGYoeSk7XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zb2xlIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWVtcHR5IjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7ImJyb3dzZXIiOnRydWUsIm5vZGUiOnRydWUsImNvbW1vbmpzIjp0cnVlLCJzaGFyZWQtbm9kZS1icm93c2VyIjp0cnVlLCJ3b3JrZXIiOnRydWUsImFtZCI6dHJ1ZSwibW9jaGEiOnRydWUsImphc21pbmUiOnRydWUsImplc3QiOnRydWUsInBoYW50b21qcyI6dHJ1ZSwianF1ZXJ5Ijp0cnVlLCJxdW5pdCI6dHJ1ZSwicHJvdG90eXBlanMiOnRydWUsInNoZWxsanMiOnRydWUsIm1ldGVvciI6dHJ1ZSwibW9uZ28iOnRydWUsInByb3RyYWN0b3IiOnRydWUsImFwcGxlc2NyaXB0Ijp0cnVlLCJuYXNob3JuIjp0cnVlLCJzZXJ2aWNld29ya2VyIjp0cnVlLCJhdG9tdGVzdCI6dHJ1ZSwiZW1iZXJ0ZXN0Ijp0cnVlLCJ3ZWJleHRlbnNpb25zIjp0cnVlLCJlczYiOnRydWUsImdyZWFzZW1vbmtleSI6dHJ1ZX19fQ==) generates no errors:
```javascript
let y = 1;
f(y);
y = f(y);
```
While [this](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZ2xvYmFscyBmICovXG5sZXQgeSA9IDE7XG55ID0gZih5KTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zb2xlIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWVtcHR5IjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7ImJyb3dzZXIiOnRydWUsIm5vZGUiOnRydWUsImNvbW1vbmpzIjp0cnVlLCJzaGFyZWQtbm9kZS1icm93c2VyIjp0cnVlLCJ3b3JrZXIiOnRydWUsImFtZCI6dHJ1ZSwibW9jaGEiOnRydWUsImphc21pbmUiOnRydWUsImplc3QiOnRydWUsInBoYW50b21qcyI6dHJ1ZSwianF1ZXJ5Ijp0cnVlLCJxdW5pdCI6dHJ1ZSwicHJvdG90eXBlanMiOnRydWUsInNoZWxsanMiOnRydWUsIm1ldGVvciI6dHJ1ZSwibW9uZ28iOnRydWUsInByb3RyYWN0b3IiOnRydWUsImFwcGxlc2NyaXB0Ijp0cnVlLCJuYXNob3JuIjp0cnVlLCJzZXJ2aWNld29ya2VyIjp0cnVlLCJhdG9tdGVzdCI6dHJ1ZSwiZW1iZXJ0ZXN0Ijp0cnVlLCJ3ZWJleHRlbnNpb25zIjp0cnVlLCJlczYiOnRydWUsImdyZWFzZW1vbmtleSI6dHJ1ZX19fQ==) does:
```javascript
let y = 1;
y = f(y);
```
And similarly [this](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZ2xvYmFscyBmICovXG5sZXQgeSA9IDE7XG5sZXQgYSA9IGYoeSk7XG55ID0gYTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zb2xlIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWVtcHR5IjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7ImJyb3dzZXIiOnRydWUsIm5vZGUiOnRydWUsImNvbW1vbmpzIjp0cnVlLCJzaGFyZWQtbm9kZS1icm93c2VyIjp0cnVlLCJ3b3JrZXIiOnRydWUsImFtZCI6dHJ1ZSwibW9jaGEiOnRydWUsImphc21pbmUiOnRydWUsImplc3QiOnRydWUsInBoYW50b21qcyI6dHJ1ZSwianF1ZXJ5Ijp0cnVlLCJxdW5pdCI6dHJ1ZSwicHJvdG90eXBlanMiOnRydWUsInNoZWxsanMiOnRydWUsIm1ldGVvciI6dHJ1ZSwibW9uZ28iOnRydWUsInByb3RyYWN0b3IiOnRydWUsImFwcGxlc2NyaXB0Ijp0cnVlLCJuYXNob3JuIjp0cnVlLCJzZXJ2aWNld29ya2VyIjp0cnVlLCJhdG9tdGVzdCI6dHJ1ZSwiZW1iZXJ0ZXN0Ijp0cnVlLCJ3ZWJleHRlbnNpb25zIjp0cnVlLCJlczYiOnRydWUsImdyZWFzZW1vbmtleSI6dHJ1ZX19fQ==) does **not**, despite being essentially the same as the previous one.
```javascript
let y = 1;
let a = f(y);
y = a;
```
Nor does [this](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZ2xvYmFscyBmICovXG5sZXQgeTtcbmYoeSlcbnkgPSA3OyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnNvbGUiOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZW1wdHkiOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnsiYnJvd3NlciI6dHJ1ZSwibm9kZSI6dHJ1ZSwiY29tbW9uanMiOnRydWUsInNoYXJlZC1ub2RlLWJyb3dzZXIiOnRydWUsIndvcmtlciI6dHJ1ZSwiYW1kIjp0cnVlLCJtb2NoYSI6dHJ1ZSwiamFzbWluZSI6dHJ1ZSwiamVzdCI6dHJ1ZSwicGhhbnRvbWpzIjp0cnVlLCJqcXVlcnkiOnRydWUsInF1bml0Ijp0cnVlLCJwcm90b3R5cGVqcyI6dHJ1ZSwic2hlbGxqcyI6dHJ1ZSwibWV0ZW9yIjp0cnVlLCJtb25nbyI6dHJ1ZSwicHJvdHJhY3RvciI6dHJ1ZSwiYXBwbGVzY3JpcHQiOnRydWUsIm5hc2hvcm4iOnRydWUsInNlcnZpY2V3b3JrZXIiOnRydWUsImF0b210ZXN0Ijp0cnVlLCJlbWJlcnRlc3QiOnRydWUsIndlYmV4dGVuc2lvbnMiOnRydWUsImVzNiI6dHJ1ZSwiZ3JlYXNlbW9ua2V5Ijp0cnVlfX19):
```javascript
let y;
f(y)
y = 1;
```

I think the [behavior](https://github.com/eslint/eslint/blob/master/lib/rules/no-unused-vars.js#L470) may in fact be (close to) what I originally stated (modifying and using on the same line doesn't count as "used", which only really makes sense for `+=`-like operators, not function parameters, though in this case the result of failing the rule is not technically incorrect), and not that it's detecting any later assignments as unused.  It seems `no-used-vars` simply checks if if a variable is *ever* assigned and *ever* used, not doing anything fancy with ordering.

**Comment by gabrieldrs** [comment]
After your first example I was wondering if it would be valid to report *all* assignements that haven't been used, what do you guys think?

**Comment by mdjermanovic** [comment]
> After your first example I was wondering if it would be valid to report _all_ assignements that haven't been used, what do you guys think?

Could you please give some examples, and would it also report some errors in a code that is now valid by this rule?

**Comment by Jimbly** [comment]
Yeah, I think reporting each assignment that is unused would be reasonable and more consistent (it's clearly what some thought was going on at the beginning of this thread).  It would, however, potentially start reporting lots of new errors in code that is now valid by this rule (e.g. all examples above would generate at least 1 error, and 3 of 5 of them do not currently generate any errors).

**Comment by gabrieldrs** [comment]
@mdjermanovic, some examples:
### Example 1
#### input:
```js
let i = 0;
i++;
```
#### current output:
```
1:5 - 'i' is assigned a value but never used.
```
#### suggested output:
```
1:5 - 'i' is assigned a value but never used.
2:1 - 'i' is assigned a value but never used.
```
---
### Example 2
#### input:
```js
/* global func */
let i = 0;
func(i);
i++;
```
#### current output:
```
```
#### suggested output:
```
4:1 - 'i' is assigned a value but never used.
```
---
### Example 3
#### input:
```js
/* global func */
let i = 0;
func(i);
i = 1;
i = 2;
i = 3;
func(i);
```
#### current output:
```
```
#### suggested output:
```
4:1 - 'i' is assigned a value but never used.
5:1 - 'i' is assigned a value but never used.
```

**Comment by Jimbly** [comment]
@gabrieldrs I think your first example should perhaps only have one error - the first assigned value is technically used by the second line, roughly equivalent to `i = i + 1`, and similar to `i = f(i)`, which definitely should count as used, though it's probably fine if a trivial increment does not count as being used (currently this rule seems to make no distinction between `i++` and `i=f(i)`).

**Comment by gabrieldrs** [comment]
You are correct, just updated it

**Comment by mdjermanovic** [comment]
That absolutely makes sense! But, it would be a huge increase in the responsibility and complexity of this rule as it would become more like a "no unused assignments" rule and most likely require some code path analysis in order to meet expectations:

```js
let i;

function f() {
   someGlobalFunction(i);
}

i = 1; // used
f(); 
i = 2; // used
f();
```

```js
let x = 1;

if (something) {
  x = 2; // unused
}
else {
  someGlobalFunction(x)
}
```

I'd suggest opening a new issue to evaluate this proposal. In the meantime, fixing just the location and/or the message in the case presented in the original post already looks like a great improvement.

**Comment by Jimbly** [comment]
Yeah, makes sense.  For the literal "no unused vars" interpretation of this rule, having glanced through the rule's code now, I'm pretty certain my original issue title and description was actually accurate - lines that both use and assign a var are ignored (great in the `i++` case, incorrect in the `i=f(i)` case).

**Comment by mdjermanovic** [comment]
> Yeah, makes sense. For the literal "no unused vars" interpretation of this rule, having glanced through the rule's code now, I'm pretty certain my original issue title and description was actually accurate - lines that both use and assign a var are ignored (great in the `i++` case, incorrect in the `i=f(i)` case).

Exactly! It reports variables that have zero "reads", while not counting those reads used to update the same variable.

**Comment by mdjermanovic** [comment]
Wondering could we just change the message if the variable is used.

```js
/* eslint no-unused-vars: error */

let x = 1;

x = f(x);
```

The actual message:

```
3:5 - 'x' is assigned a value but never used. (no-unused-vars)
```

This message is technically incorrect, so changing only the text for these cases might be a bug fix.

**Comment by kaicataldo** [comment]
I'm always :+1: for better error messages! Would you like to make a new issue?

**Comment by Jimbly** [comment]
The inherent problem is the code is actually detecting the variable as completely unused (it's treating `x = f(x)` as syntactically equivalent to `x++`), I think it might just be better to fix the code to consider that "used" (error message would completely go away, just like in `let x=1; f(x); x = f(x);`.  The most "correct' error message when your code looks like `let x = 1; x=x+1;` would be something like `'x' is assigned a value but only ever assigned back to itself`, but that's not even correct in the `x=f(x)` case.  If you can figure out how to differentiate those cases inside of the [`isSelfReference`](https://github.com/eslint/eslint/blob/fd8e1f52110cada542a120750236fd1ec8779336/lib/rules/no-unused-vars.js#L496) logic in order to display a different error message, you could just fix the bug =)
